### PR TITLE
Fixed copy internal-link into new language

### LINKED ISSUE #2313: Page translation for page with type "internal link"
| Q | A
| --- | ---
| Bug? | yes
| New Feature? | no
| Sulu Version | 1.2.0
| Browser Version | All

#### Actual Behavior
Adding a translation for a page with "internal  link" as page type does not work as expected:

When selecting "Create a new empty page in this language": 
- A page with type internal link is created
- The content of "Linked content" is copied and "No Title (/main/sub-1)" is shown (even if the linked page is not available in this language)
- Selecting "content" as page type and saving afterwords leads to an error ("Document has no title (title is required for auto name behavior)")

internal-link

When selecting "Create a new page in this language and copy content from": 
- The content of "Linked content" is copied and "No Title (/main/sub-1)" is shown (even if the linked page is not available in this language)

#### Expected Behavior
When selecting "Create a new empty page in this language": 
- A page with type content should be created.

When selecting "Create a new page in this language and copy content from": 
- The content of "Linked content" should not be copied if the selected page is not available in the current language.

### PR DESCRIPTION
**Pull request by wachterjohannes**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #2313
| Related issues/PRs | https://github.com/sulu/sulu-document-manager/pull/82 and https://github.com/massiveart/husky/pull/650
| License | MIT
| Documentation PR | none

#### What's in this PR?

This PR updates sulu-document-manager and additionally fixes some javascript issues with copy internal link content to a new language.

#### Why?

There were several issues with copy internal-link pages.

### CODE REVIEW

**Comment by chirimoya** [comment]
tmp ???

**Comment by wachterjohannes** [comment]
@chirimoya i have updated the title (: but that was only for me to see which tests fail (:

File: composer.json [hunk 1]
```diff
@@ -25,7 +25,7 @@
         "imagine/imagine": "~0.6.1",
         "liip/theme-bundle": "1.3.*",
         "massive/search-bundle": "0.13.*",
-        "sulu/document-manager": "dev-master",
+        "sulu/document-manager": "dev-enhancement/document-registry",
```
Comment (wachterjohannes) [inline composer.json]: have to be remove before merge

File: src/Sulu/Bundle/ContentBundle/Tests/Features/content_management.feature [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        #  - drop-down item should have a speaking name
+        And I click "0" from the drop down
+        And I confirm
+        And I wait and expect to see element "#title"
+        Then I expect the value of the property "title" is "Behat Test Content"
+
+# TODO missing testcases
+#  - internal link
+#  - external link
+#  - multiple step tests (create - translate - to list)
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Tests/Features/content_management.feature]: we have currently some really big problems with routing in the behat tests. after one routing the events will be broken and for example the save-button wont be triggered and cannot be clicked.

File: src/Sulu/Bundle/AdminBundle/Behat/AdminContext.php [hunk 3]
```diff
@@ -57,6 +57,14 @@ public function iExpectAConfirmationDialogShouldAppear()
     }
 
     /**
+     * @Given I click the back icon
+     */
+    public function iClickOnTheBackIcon()
+    {
+        $this->clickSelector('.fa-chevron-left');
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Behat/AdminContext.php]: This is currently not working and therefore not being used, right? Then I would remove it for now (since it might have to be solved in a different way anyway)

File: src/Sulu/Bundle/ContentBundle/Serializer/Subscriber/StructureSubscriber.php [hunk 4]
```diff
@@ -113,16 +113,13 @@ private function addStructureProperties(StructureMetadata $structureMetadata, St
     {
         /** @var ManagedStructure $structure */
         $structure = $document->getStructure();
-        foreach ($structure->toArray() as $name => $value) {
-            if ($name === 'title') {
+        $data = $structure->toArray();
+        foreach ($structureMetadata->getProperties() as $name => $property) {
+            if ($name === 'title' || !array_key_exists($name, $data) || $property->hasTag('sulu.rlp')) {
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Serializer/Subscriber/StructureSubscriber.php]: Why is that change required?

File: src/Sulu/Bundle/SnippetBundle/Document/SnippetDocument.php [hunk 5]
```diff
@@ -205,6 +206,22 @@ public function setLocale($locale)
...
+        return $this->originalLocale;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setOriginalLocale($originalLocale)
+    {
+        $this->originalLocale = $originalLocale;
+    }
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Document/SnippetDocument.php]: That's another implication of https://github.com/sulu/sulu-document-manager/pull/82/files#r63675487. The snippet (at least currently) always has only one language, so ther is no need to distinguish between original locale and locale.

File: src/Sulu/Component/Content/Document/Subscriber/ResourceSegmentSubscriber.php [hunk 6]
```diff
@@ -100,11 +100,12 @@ public function handleHydrate(AbstractMappingEvent $event)
 
         $node = $event->getNode();
         $property = $this->getResourceSegmentProperty($document);
-        $originalLocale = $this->documentInspector->getOriginalLocale($document);
+        $locale = $this->documentInspector->getOriginalLocale($document);
+        $locale = $this->documentInspector->getOriginalLocale($document);
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Subscriber/ResourceSegmentSubscriber.php]: One time is probably enough 😉

File: src/Sulu/Component/Content/Mapper/ContentMapper.php [hunk 7]
```diff
@@ -725,22 +729,22 @@ private function copyOrMove($uuid, $destParentUuid, $userId, $webspaceKey, $loca
             $this->documentManager->refresh($parentDocument);
         }
 
+        $originalDocument = $document;
```
Comment (danrot) [inline src/Sulu/Component/Content/Mapper/ContentMapper.php]: Why not naming the upper `$document` `$originalDocument` immediately? Then we would not have to reassign something to `$document`.

File: src/Sulu/Component/Content/Mapper/ContentMapper.php [hunk 8]
```diff
@@ -770,7 +774,7 @@ private function copyOrMove($uuid, $destParentUuid, $userId, $webspaceKey, $loca
 
         $this->documentManager->find($document->getUuid(), $originalLocale);
 
-        return $this->documentToStructure($document);
+        return $this->documentToStructure($originalDocument);
```
Comment (danrot) [inline src/Sulu/Component/Content/Mapper/ContentMapper.php]: I am not sure about this, but wouldn't it make more sense to return the new document here?

File: src/Sulu/Component/Content/Tests/Functional/Mapper/ContentMapperTest.php [hunk 9]
```diff
@@ -2141,9 +2138,7 @@ public function testCopyLanguageTree()
     public function testNodeAndAncestorsExcludedGhosts()
     {
         $data = $this->prepareBigTreeTestData();
-        $child = $data[1]->getChildren()[0]->getChildren()[2]->getChildren()[1];
-
-        $result = $this->mapper->loadNodeAndAncestors($child->getUuid(), 'en', 'sulu_io', true);
+        $result = $this->mapper->loadNodeAndAncestors($data[8]->getUuid(), 'en', 'sulu_io', true);
```
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Functional/Mapper/ContentMapperTest.php]: Why has the structure changed here?

File: src/Sulu/Bundle/AdminBundle/Behat/AdminContext.php (hunk 3 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Behat/AdminContext.php]: it is used but on the page will redirect no events will work. but i use it to check the save (see feature files).

File: src/Sulu/Component/Content/Tests/Unit/Document/Subscriber/FallbackLocalizationSubscriberTest.php [hunk 10]
```diff
@@ -132,14 +128,9 @@ public function testNoWebspaceDoNotLoadGhostContent()
     {
         $this->inspector->getWebspace($this->document->reveal())->willReturn(null);
         $this->inspector->getLocales($this->document)->willReturn(['de', 'fr']);
-        $this->hydrateEvent->setLocale('de')->shouldBeCalled();
+        $this->hydrateEvent->setLocale('de')->shouldNotBeCalled();
```
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Unit/Document/Subscriber/FallbackLocalizationSubscriberTest.php]: I get that `updateLocale` shouldn't be called anymore, but shouldn't it still change the localization of the event, so that the next subscribers take the right language?

File: src/Sulu/Bundle/SnippetBundle/Document/SnippetDocument.php (hunk 5 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/SnippetBundle/Document/SnippetDocument.php]: see the discussion i have added a comment there.

File: src/Sulu/Component/Content/Mapper/ContentMapper.php (hunk 8 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Content/Mapper/ContentMapper.php]: no also before the "original" was returned. that was done by calling find with the original locale right before this line.

File: src/Sulu/Component/Content/Tests/Functional/Mapper/ContentMapperTest.php (hunk 9 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Content/Tests/Functional/Mapper/ContentMapperTest.php]: because its more readable in my opinion and i would have to reload the documents because of the proxy generation in the hydrate event.

File: src/Sulu/Component/Content/Tests/Unit/Document/Subscriber/FallbackLocalizationSubscriberTest.php (hunk 10 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Content/Tests/Unit/Document/Subscriber/FallbackLocalizationSubscriberTest.php]: no because if you dont load ghost content it should not be changed for other event subscribers.

File: src/Sulu/Bundle/ContentBundle/Serializer/Subscriber/StructureSubscriber.php (hunk 4 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Serializer/Subscriber/StructureSubscriber.php]: because in some reasons you get not a managed Structure (which only returns values for existing properties). you get a structure which also returns for example shadowOn as a structure property. which is actually not but the mapping create a structure property-value for it. to be sure that such values are not published over the api this logic was introduced.

**Comment by wachterjohannes** [comment]
closed in favor of https://github.com/sulu/sulu/pull/2405
